core/jvm/src/main/scala/zio/internal/ZScheduler.scala

```diff
@@ -200,74 +198,58 @@ private final class ZScheduler(autoBlocking: Boolean) extends Executor {
         } else {
-          globalQueue.offerAll(worker.localQueue.pollUpTo(128) :+ runnable)
+          val rnd = ThreadLocalRandom.current()
+          globalQueue.offerAll(worker.localQueue.pollUpTo(128), rnd)
+          globalQueue.offer(runnable, rnd)
           notify = true
         }
       } else {
         globalQueue.offer(runnable)
         notify = true
       }
       if (notify) {
-        val currentState     = state.get
-        val currentActive    = (currentState & 0xffff0000) >> 16
-        val currentSearching = currentState & 0xffff
-        if (currentActive != poolSize && currentSearching == 0) {
-          var loop = true
-          while (loop) {
-            val worker = idle.poll(null)
-            if (worker eq null) {
-              loop = false
-            } else {
-              state.getAndAdd(0x10001)
-              worker.active = true
-              LockSupport.unpark(worker)
-              loop = false
-            }
-          }
-        }
+        val currentState = state.get
+        maybeUnparkWorker(currentState)
       }
       true
     }
+  }
 
-  private[this] def isBlocking(runnable: Runnable): Boolean =
-    if (runnable.isInstanceOf[FiberRunnable]) {
+  private[this] def isBlocking(worker: ZScheduler.Worker, runnable: Runnable): Boolean =
+    if (autoBlocking && runnable.isInstanceOf[FiberRunnable]) {
       val fiberRunnable = runnable.asInstanceOf[FiberRunnable]
       val location      = fiberRunnable.location
-      submittedLocations.put(location)
-      blockingLocations.contains(location)
-    } else {
-      false
-    }
+      if ((location ne null) && (location ne emptyTrace)) {
+        if (worker eq null) globalLocations.put(location)
+        else worker.submittedLocations.put(location)
+        blockingLocations.contains(location)
+      } else false
+    } else false
 
   private[this] def makeLocations(): ZScheduler.Locations =
-    new ZScheduler.Locations {
-      private[this] val locations = new java.util.HashMap[Trace, Array[Long]]
-      def get(trace: Trace): Long = {
-        val array = locations.get(trace)
-        if (array eq null) 0L else array(0)
-      }
-      def put(trace: Trace): Long = {
-        val array = locations.get(trace)
-        if (array eq null) {
-          locations.put(trace, Array(1L))
-          0L
-        } else {
-          val value = array(0)
-          array(0) += 1
-          value
-        }
-      }
-    }
+    if (autoBlocking) new ZScheduler.Locations.Enabled
+    else ZScheduler.Locations.Disabled
 
   private[this] def makeSupervisor(): ZScheduler.Supervisor =
     new ZScheduler.Supervisor {
+
+      private def countSubmittedAt(location: Trace): Long = {
+        var count = globalLocations.get(location)
+        var i     = 0
+        while (i < poolSize) {
+          val workerCount = workers(i).submittedLocations.get(location)
+          count += workerCount
+          i += 1
+        }
+        count
+      }
```

Comment:
Here we know that if `autoBlocking` is false, all the calls to `workers(i).submittedLocations.get(location)` will return `0` so we iterate over the workers for nothing.

It's duplicating the logic but we could avoid these iterations if we write:
```scala
      private def countSubmittedAt(location: Trace): Long =
        if (autoBlocking) {
          var count = globalLocations.get(location)
          var i     = 0
          while (i < poolSize) {
            val workerCount = workers(i).submittedLocations.get(location)
            count += workerCount
            i += 1
          }
          count
        } else globalLocations.get(location)
```

WDYT?